### ISSUE TITLE
TaggedImageProcessor and Acquisition

### ISSUE THREAD

**Issue by jdeschamps** [micro-manager/AcqEngJ#47]
Hi!

I have long wanted to switch our acquisition plugin from the standard MM acquisitions to AcqEngJ. However, I am having a hard time figuring out how to use a `TaggedImageProcessor` correctly. Let's say I have a set-up like this:

``` Java
Engine engine = new Engine(studio.core());
	
// create sink and acquisition
SimpleSink sink = new SimpleSink(); // just counts images
Acquisition acq = new Acquisition(sink);
		
// create processor
SimpleProcessor q = new SimpleProcessor();
		
// create acquisition event list
ArrayList<AcquisitionEvent> list = new ArrayList<AcquisitionEvent>();
for(int i=0; i<10; i++) {
	AcquisitionEvent a = new AcquisitionEvent(acq);
	a.setExposure(50);
	a.setTimeIndex(i);
	list.add(a);
}

// start acquisition?
acq.submitEventIterator(list.iterator());
		
// adding processor after the acquisition has started leads to missing frames in the processor
acq.addImageProcessor(q);
		
// finish
acq.finish();
```
The `SimpleProcessor.setDequeues` looks like this:

``` Java
int counter  = 0;
while(true) {			
	TaggedImage newImage = source.pollFirst();
	if(newImage != null) {
		if (newImage.tags == null && newImage.pix == null) {
			System.out.println("[proc] done");
			break;
		}
			
		sink.add(newImage);
		System.out.println("[proc] pass on image "+counter++);
	}
}
```

Since `acq.addImageProcessor` calls `setDequeues`, calling it before `acq.submitEventIterator` causes the `while` loop to block the thread. If I call `acq.addImageProcessor` after `acq.submitEventIterator` then I miss frames.

So a bunch of questions:
- Have I been doing things wrong? And are there examples somewhere of correct implementations?
- Would it make sense to only instantiate the queues/deques in `Acquisition.addImageProcessor` and to put the calls to `TaggedImageProcessor.setDequeues` before running events?
- From an API point of view, I would not expect `submitEventIterator` to start iterating over the events right away. I guess that `Acquisition` is designed so that events can be added any time and there can be pauses between batch of events. But a global switch (e.g. `start()`) could be useful in that case to set up processors, while not changing the behaviour afterwards. Not sure if what I mean is clear!

Thanks!!

**Comment by henrypinkard** [comment]
Yes, it all makes sense. The only example IamgeProcessor thus far is  [RemoteImageProccessor](https://github.com/micro-manager/pycro-manager/blob/main/java/src/main/java/org/micromanager/remote/RemoteImageProcessor.java), which has two threads, one which pushes images out across a socket and one that pulls them back in (thereby allowing image processing to be done in python/pycromanager).

The fix with the current API for you would be to run your image processor on a different thread, and add it before submitting the event iterator. Like this:

```java
ExecutorService es = Executors.newSingleThreadExecutor(
              (Runnable r) -> new Thread(r, "ImageProcessorThread"));

es.submit(() -> {
int counter  = 0;
while(true) {			
	TaggedImage newImage = source.pollFirst();
	if(newImage != null) {
		if (newImage.tags == null && newImage.pix == null) {
			System.out.println("[proc] done");
                        es.shutdown();
			break;
		}
			
		sink.add(newImage);
		System.out.println("[proc] pass on image "+counter++);
	}
}

        });
```

> From an API point of view, I would not expect submitEventIterator to start iterating over the events right away. I guess that Acquisition is designed so that events can be added any time and there can be pauses between batch of events. But a global switch (e.g. start()) could be useful in that case to set up processors, while not changing the behaviour afterwards

It's a good point. At the very least I should document this. There's actually an internal [`start()`](https://github.com/henrypinkard/AcqEngJ/blob/b3e332e3fa1a8ecab4f2b2b518088a2768c544eb/src/main/java/org/micromanager/acqj/main/Acquisition.java#L153) already so it simple to break this into the API.

Though if you were to process images on the same thread as in your example above you still have an ugly situation where you'd have to submit events from your `ImageProcessor`, which is not totally straightforward. 

So maybe the solution here is just to document what I've just explained in the README? Curious to know what you think.

Also, I'm currently still making some small tweaks to the API so that MMStudio can call it, but it should be pretty easy to update your code if thats needed

**Comment by jdeschamps** [comment]
Thanks for the tip! I had to add a `Thread.sleep()` before adding the events to avoid losing the first frame.

> It's a good point. At the very least I should document this. There's actually an internal [start()](https://github.com/henrypinkard/AcqEngJ/blob/b3e332e3fa1a8ecab4f2b2b518088a2768c544eb/src/main/java/org/micromanager/acqj/main/Acquisition.java#L153) already so it simple to break this into the API.
>
>Though if you were to process images on the same thread as in your example above you still have an ugly situation where you'd have to submit events from your ImageProcessor, which is not totally straightforward.

What we do is currently simple: the ImageProcessor is just there to monitor the number of peaks per images, so right now it does what I needl. If we were to do more involved feedback-loop, this could become tricky indeed.  

Naively I would have expected an API with `start()`/`abort()`/`pause()`, `addProcessor()`/`removeProcessor()` and `insertEvents()`/`clearEvents()`. That way the API would indicate that we could start a series of event, pause the acquistion when an `ImageProcessor` detects something, clear the events and submit a new batch depending on the `ImageProcessor` state. But then that's because I thought of the `Acquisition` as an all-encompassing acquisition engine. 

I guess the better use of it in case of feedback-loop is to create multiple `Acquisition` instances, for instance for a z-stack on detected events:
- overall loop on stage positions (main thread)
- `acq1` acquires a time series with `imProc1` detecting a particular event (on a different thread) and sending the frames to `sink1`
- when an event is detected by `imProc1`, flag goes up (e.g. atomic boolean) and in the main thread `acq1` is aborted
- `acq2` is created and for instance a list of events corresponding to a z-stack is submitted 
- once the z-stack is acquired, move to next position and start again.

In such a case, then the only thing is to make dev/users aware of the indented scope of `Acquisition`.

> So maybe the solution here is just to document what I've just explained in the README? Curious to know what you think.

Not only in the Readme but also in the docstring of `Acquisition`, `AcquisitionAPI` and `ImageProcessor`, and by adding a toy example to the repo (this one could live on the Readme, but I would opt for an example package, since probably more examples would come in the future).

Basically, having had a deeper look at the code, I now realise that `start()` is called directly when creating the `Acquisition`. I think this should be clarified in the class and constructor docstring!  
 
I think once the documentation is bit more explicit about the usage, it is going to be a lot easier to adopt in third-party plugins! 

> Also, I'm currently still making some small tweaks to the API so that MMStudio can call it, but it should be pretty easy to update your code if thats needed

Yes I expected that! I am currently also just exploring how I could convert our acquisitions, so I'll keep an eye on the changes.

**Comment by henrypinkard** [comment]
Thanks for the feedback

> Thanks for the tip! I had to add a `Thread.sleep()` before adding the events to avoid losing the first frame.

I'm unclear if you've moved away from this strategy, but I just want to emphasize that this is a bad idea because there is no guarantee that that sleep will have been long enough to ensure that the image processor gets inserted in time.

> Naively I would have expected an API with `start()`/`abort()`/`pause()`, `addProcessor()`/`removeProcessor()` and `insertEvents()`/`clearEvents()`. That way the API would indicate that we could start a series of event, pause the acquistion when an `ImageProcessor` detects something, clear the events and submit a new batch depending on the `ImageProcessor` state. But then that's because I thought of the `Acquisition` as an all-encompassing acquisition engine.

I think all this functionality is available with current API, with the exception of `start()`. Since `submitEventIterator` returns a `Future`, it can be canceled and the events from the iterator ceased


> But then that's because I thought of the `Acquisition` as an all-encompassing acquisition engine.

I'm unclear on what this means. Could you explain?

> I guess the better use of it in case of feedback-loop is to create multiple `Acquisition` instances, for instance for a z-stack on detected events:

This might work (though untested as far as I know). But I don't think this is necessary since you can call `submitEventIterator` multiple times and cancel as needed.

> Not only in the Readme but also in the docstring of `Acquisition`, `AcquisitionAPI` and `ImageProcessor`, and by adding a toy example to the repo (this one could live on the Readme, but I would opt for an example package, since probably more examples would come in the future).
> 
> Basically, having had a deeper look at the code, I now realise that `start()` is called directly when creating the `Acquisition`. I think this should be clarified in the class and constructor docstring!
> 
> I think once the documentation is bit more explicit about the usage, it is going to be a lot easier to adopt in third-party plugins!

Thanks for the suggestions! I will definitely add this.

**Comment by henrypinkard** [comment]
Okay I've tried to implement your many good suggestions. I've made an [example](https://github.com/micro-manager/AcqEngJ/tree/main/src/main/java/org/micromanager/acqj/example) folder that shows how to build this library. Closing for now, but feel free to reopen or open another if you have more questions/suggestions

**Comment by jdeschamps** [comment]
Thanks a lot! 

Sorry for the lack of answer, got caught up in other things and lost track of this.

> > But then that's because I thought of the `Acquisition` as an all-encompassing acquisition engine.
> 
> I'm unclear on what this means. Could you explain?

I am also confused at what I meant. I guess at first I approached the `Acquisition` thinking that it had full control of the acquisition but also the processing, in a single-thread manner. Here the processing and the acquisition are decoupled But users/developers need to be careful in synchronizing the threads. Now it's too far back so I am not sure what my thread of thought was. :)

Hopefully I'll have time soon to go back to experimenting with the library!
